Allow bots to provide client cert authentication when rejoining

timothyb89 opened the pull request:
Traditionally, bots using non-`token` join methods "renew" their certs by rejoining, taking advantage of their repeatable join methods to avoid relying on `token` rejoin logic. However, as part of [RFD 162], we need a way for rejoining bots to assert an existing identity (if any) even when rejoining "from scratch".

When paired with #43732, this change allows bots to use an authenticated auth client for all "initial" joins. If present, an existing `BotInstanceID` is extracted from the client identity and used to associate the bot with its previous identity and add a new authentication record to its `BotInstance`.

As certain join methods (TPM, Azure) are initially handled by the `JoinService` using gRPC, this additionally adds a new `BotInstanceID` field to `RegisterUsingTokenRequest` so the `JoinService` can verify the client identity and extract the `BotInstanceID` if available. Like `RemoteAddr`, this value is only trusted when set by the proxy and should be ignored in all other cases.

[RFD 162]: https://github.com/gravitational/teleport/pull/36510/

timothyb89 committed:
Allow bots to provide client cert authentication when rejoining

Traditionally, bots using non-`token` join methods "renew" their
certs by rejoining, taking advantage of their repeatable join methods
to avoid relying on `token` rejoin logic. However, as part of
[RFD 162], we need a way for rejoining bots to assert an existing
identity (if any) even when rejoining "from scratch".

When paired with #43732, this change allows bots to use an
authenticated auth client for all "initial" joins. If present, an
existing `BotInstanceID` is extracted from the client identity and
used to associate the bot with its previous identity.

As certain join methods (TPM, Azure) are initially handled by the
`JoinService` using gRPC, this additionally adds a new
`BotInstanceID` field to `RegisterUsingTokenRequest` so the
`JoinService` can verify the client identity and extract the
`BotInstanceID` if available. Like `RemoteAddr`, this value is only
trusted when set by the proxy and should be ignored in all other
cases.

[RFD 162]: https://github.com/gravitational/teleport/pull/36510/

timothyb89 commented:
This change touches a lot of join logic and I think exercises a not-often-used path (join via authenticated auth client) so I want to test more thoroughly than normal. 

Unit tests are WIP still, so this is drafted until (at least) then.

strideynet commented on lib/auth/bot.go:
```diff
@@ -392,7 +383,41 @@ func (a *Server) generateInitialBotCerts(
+
+			// Note: botName is derived from the provision token rather than any
+			// value sent by the client, so we can trust it.
+			// TODO: Determine how we should handle update failure in edge cases
```
I feel like we have two options here:

- Fail hard on the RPC side and mitigate this failure within `tbot`. This sounds a little more complex than I'd first think since `tbot` would have to create a new client w/o it's credentials and then rejoin.
- Create a new bot instance and then return certificates with that bot instance ID. 

I think generally, I like the idea of handling this server-side by just returning a new bot instance (as if none had been specified at all). If we design `tbot` around the idea that it's bot instance ID can change at the auth server's whim, then we shouldn't have a problem. I think this also nicely handles scenarios where someone has torn down their cluster, and restored it via Terraform or similar, the bot/token would exist, but the bot instances wouldn't. WDYT?

timothyb89 replied:
I'm generally inclined to agree, I think the critical security control is that we don't ever reissue an instance ID unless we're certain it's the same bot; this wouldn't impact that at all, so I think it's fine.

I've added some logic to create a new instance here if the requested ID isn't found, and added a testcase for this behavior.

strideynet reviewed (approved):
Generally looks good - I think it'd be beneficial to have a few tests here though!

timothyb89 committed:
Add tests for authenticated bot rejoining

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-instance-reauthentication

timothyb89 commented:
I've done a pretty thorough manual testing pass, including deploying this on a Cloud tenant with k8s and TPM auth; all seems well.

The tests came out a bit more, eh, cursed than I'd hoped since we need to authenticate through the "front door", so to speak, to properly exercise the proxy's `BotInstanceID` parameter trust checks (with `RegisterUsingTokenRequest`). I've had to clumsily pull in the [`join.go` changes](https://github.com/gravitational/teleport/pull/43732/files#diff-42f6948b785d320a3a1164dcb245ef4a3e24aad457b82525523991d5cac8a78f) from #43732 so we could use `join.Register()` with a preexisting client, so we should probably merge that first and I can remove the duplicate changes.

(I just ended up copying the content of that single file because rebasing onto that PR's branch proved to be... bad. I'll just deal with the merge conflict consequences here once it's merged.)

timothyb89 committed:
Create a new bot instance if the requested instance is not found

Also, includes a new check in for this behavior in
`TestRegisterBotWithInvalidInstanceID`.

timothyb89 committed:
Fix imports

zmb3 commented on lib/auth/auth_with_roles.go:
```diff
@@ -574,13 +574,33 @@ func (a *ServerWithRoles) GetClusterCACert(
+	// If the identity has a BotInstanceID included, copy it onto the request -
+	// but only if one wasn't already passed along via the proxy.
+	ident := a.context.Identity.GetIdentity()
+	if req.BotInstanceID == "" && ident.BotInstanceID != "" {
+		req.BotInstanceID = ident.BotInstanceID
+	}
```
```suggestion
	req.BotInstanceID = cmp.Or(req.BotInstanceID, ident.BotInstanceID)
```

zmb3 reviewed (commented):
Looks good though I agree with Noah that we should have better test coverage.

timothyb89 committed:
Update lib/auth/auth_with_roles.go

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-instance-reauthentication

timothyb89 committed:
Fix `cmp` import

timothyb89 commented:
> Looks good though I agree with Noah that we should have better test coverage.

Yeah, as Noah said, I'm absolutely open to suggestions for additional coverage, but the existing `TestRegisterBot_BotInstanceRejoin` and `TestRegisterBotWithInvalidInstanceID` tests cover all the critical paths I'm most concerned about. Coverage of all individual join methods would be ideal but may not be practical, and I'm hopeful we've reached an acceptable compromise with both gRPC and HTTPS methods represented (IAM and Kubernetes, plus `token`).
